File: artifacts/slack-clone/trial-1/source/app/irc.py
"""Minimal IRC gateway bridging an authenticated workspace into IRC.

Each TCP connection presents a single user (selected by `PASS <token>`).
Joins map to channels in the user's first owned/membered workspace whose name
matches; PRIVMSG into the same per-channel `seq` stream as the HTTP API.
Web messages fan out to IRC clients via the same DB-tail loop.
"""

from __future__ import annotations

import asyncio
import json
import os
import time
from typing import Optional

from . import db, util


IRC_HOST = os.environ.get("HUDDLE_IRC_HOST", "0.0.0.0")
IRC_PORT = int(os.environ.get("HUDDLE_IRC_PORT", "6667"))
SERVER_NAME = "huddle.irc"


# --- Helpers ---------------------------------------------------------------

def _send(writer: asyncio.StreamWriter, line: str):
    if not line.endswith("\r\n"):
        line += "\r\n"
    try:
        writer.write(line.encode("utf-8", errors="replace"))
    except Exception:
        pass


def _user_by_token(token: str) -> Optional[dict]:
    row = db.conn().execute(
        "SELECT u.* FROM tokens t JOIN users u ON u.id = t.user_id WHERE t.token = ?",
        (token,),
    ).fetchone()
    return dict(row) if row else None


def _user_workspaces(user_id: str) -> list[dict]:
    rows = db.conn().execute(
        "SELECT w.* FROM workspaces w JOIN workspace_members m ON m.workspace_id = w.id "
        "WHERE m.user_id = ? ORDER BY w.created_at ASC",
        (user_id,),
    ).fetchall()
    return [dict(r) for r in rows]


def _channel_by_name(workspace_id: str, name: str) -> Optional[dict]:
    name = name.lstrip("#").lower()
    row = db.conn().execute(
        "SELECT * FROM channels WHERE workspace_id = ? AND name = ?",
        (workspace_id, name),
    ).fetchone()
    return dict(row) if row else None


# --- Session ---------------------------------------------------------------

class IrcSession:
    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter,
                 server: "IrcGateway"):
        self.reader = reader
        self.writer = writer
        self.server = server
        peer = writer.get_extra_info("peername")
        self.host = (peer[0] if peer else "irc")
        self.token: Optional[str] = None
        self.user: Optional[dict] = None
        self.nick: str = ""
        self.user_set: bool = False
        self.registered: bool = False
        self.workspace: Optional[dict] = None
        self.joined: dict[str, dict] = {}  # name -> channel row
        self._closed = False

    def prefix(self) -> str:
        return f":{self.nick}!{(self.user['username'] if self.user else 'h')}@{self.host}"

    def send(self, line: str):
        _send(self.writer, line)

    def numeric(self, code: str, args: str):
        self.send(f":{SERVER_NAME} {code} {self.nick or '*'} {args}")

    async def run(self):
        self.send(f":{SERVER_NAME} NOTICE * :*** Huddle IRC gateway")
        try:
            while not self._closed:
                line = await self.reader.readline()
                if not line:
                    break
                line = line.decode("utf-8", errors="replace").rstrip("\r\n")
                if not line:
                    continue
                await self.handle_line(line)
        except Exception:
            pass
        finally:
            self.server.remove(self)
            try:
                self.writer.close()
                await self.writer.wait_closed()
            except Exception:
                pass

    async def handle_line(self, line: str):
        # Parse: optional ":prefix" then COMMAND args... [:trailing]
        if line.startswith(":"):
            _, _, rest = line.partition(" ")
            line = rest
        cmd = ""
        args: list[str] = []
        rest = line
        if " :" in rest:
            head, _, trailing = rest.partition(" :")
            parts = head.split()
            args = parts[1:] if parts else []
            cmd = parts[0].upper() if parts else ""
            args.append(trailing)
        else:
            parts = rest.split()
            if parts:
                cmd = parts[0].upper()
                args = parts[1:]
        if not cmd:
            return
        if cmd == "PASS":
            self.token = args[0] if args else ""
            return
        if cmd == "NICK":
            new_nick = args[0] if args else ""
            if not self.token:
                self.numeric("464", ":Password required")
                return
            if not self.user:
                u = _user_by_token(self.token)
                if not u:
                    self.numeric("464", ":Bad password")
                    return
                self.user = u
                ws_list = _user_workspaces(u["id"])
                if ws_list:
                    self.workspace = ws_list[0]
            # Detect collision: existing session with same nick on same workspace
            for s in self.server.sessions:
                if s is self:
                    continue
                if s.nick.lower() == new_nick.lower() and s.workspace and self.workspace and \
                        s.workspace["id"] == self.workspace["id"]:
                    self.numeric("433", f"{new_nick} :Nickname is already in use.")
                    return
            self.nick = new_nick
            await self._maybe_register()
            return
        if cmd == "USER":
            self.user_set = True
            await self._maybe_register()
            return
        if not self.registered:
            # Most commands require registration.
            if cmd not in ("CAP", "QUIT", "PASS"):
                self.numeric("451", ":You have not registered")
                return
        if cmd == "CAP":
            # No capability negotiation; ack req with NAK so registration can proceed.
            sub = args[0].upper() if args else ""
            if sub == "LS":
                self.send(f":{SERVER_NAME} CAP * LS :")
            elif sub == "REQ":
                want = args[1] if len(args) > 1 else ""
                self.send(f":{SERVER_NAME} CAP * NAK :{want}")
            elif sub == "END":
                pass
            return
        if cmd == "PING":
            payload = args[0] if args else SERVER_NAME
            self.send(f":{SERVER_NAME} PONG {SERVER_NAME} :{payload}")
            return
        if cmd == "PONG":
            return
        if cmd == "QUIT":
            self._closed = True
            return
        if cmd == "JOIN":
            if not args:
                return
            for chan_arg in args[0].split(","):
                await self._join(chan_arg)
            return
        if cmd == "PART":
            if not args:
                return
            for chan_arg in args[0].split(","):
                self._part(chan_arg)
            return
        if cmd == "NAMES":
            if not args:
                return
            for chan_arg in args[0].split(","):
                self._names(chan_arg)
            return
        if cmd == "WHO":
            target = args[0] if args else ""
            self.numeric("315", f"{target} :End of /WHO list")
            return
        if cmd == "LIST":
            self._list()
            return
        if cmd == "TOPIC":
            if not args:
                return
            chan = args[0]
            new_topic = args[1] if len(args) > 1 else None
            await self._topic(chan, new_topic)
            return
        if cmd == "MODE":
            target = args[0] if args else ""
            if target.startswith("#"):
                self.numeric("324", f"{target} +n")
            return
        if cmd == "PRIVMSG":
            if len(args) < 2:
                return
            await self._privmsg(args[0], args[1])
            return
        # Unknown
        self.numeric("421", f"{cmd} :Unknown command")

    async def _maybe_register(self):
        if self.registered:
            return
        if not self.user or not self.nick or not self.user_set:
            return
        self.registered = True
        ws_label = (self.workspace["name"] if self.workspace else "Huddle")
        self.numeric("001", f":Welcome to {ws_label}, {self.nick}")
        self.numeric("002", f":Your host is {SERVER_NAME}")
        self.numeric("003", ":This server was created today")
        self.numeric("004", f"{SERVER_NAME} 1.0 i tnp")
        self.numeric("005", f"PREFIX=(o)@ CHANTYPES=# CASEMAPPING=ascii :are supported by this server")
        self.numeric("375", f":- {SERVER_NAME} Message of the day -")
        self.numeric("372", f":- Welcome to the Huddle IRC bridge.")
        self.numeric("376", ":End of /MOTD command")
        self.server.add(self)

    async def _join(self, chan_name: str):
        if not chan_name.startswith("#"):
            self.numeric("403", f"{chan_name} :No such channel")
            return
        if not self.workspace:
            self.numeric("403", f"{chan_name} :No such channel")
            return
        ch = _channel_by_name(self.workspace["id"], chan_name)
        if not ch:
            self.numeric("403", f"{chan_name} :No such channel")
            return
        # Add channel membership in DB.
        try:
            db.conn().execute("BEGIN IMMEDIATE")
            db.conn().execute(
                "INSERT OR IGNORE INTO channel_members(channel_id, user_id, joined_at, last_read_seq) "
                "VALUES (?, ?, ?, 0)",
                (ch["id"], self.user["id"], db.now_ms()),
            )
            db.conn().execute("COMMIT")
        except Exception:
            db.tx_rollback()
        self.joined[ch["name"]] = ch
        # Send JOIN, TOPIC, NAMES.
        self.send(f"{self.prefix()} JOIN :#{ch['name']}")
        if ch["topic"]:
            self.numeric("332", f"#{ch['name']} :{ch['topic']}")
        else:
            self.numeric("331", f"#{ch['name']} :No topic is set")
        names = []
        for r in db.conn().execute(
            "SELECT u.username FROM channel_members cm JOIN users u ON u.id = cm.user_id "
            "WHERE cm.channel_id = ?",
            (ch["id"],),
        ).fetchall():
            names.append(r["username"])
        self.numeric("353", f"= #{ch['name']} :{' '.join(names)}")
        self.numeric("366", f"#{ch['name']} :End of NAMES")

    def _part(self, chan_name: str):
        name = chan_name.lstrip("#").lower()
        if name in self.joined:
            self.send(f"{self.prefix()} PART #{name}")
            del self.joined[name]

    def _names(self, chan_name: str):
        name = chan_name.lstrip("#").lower()
        ch = self.joined.get(name) or (self.workspace and _channel_by_name(self.workspace["id"], name))
        if not ch:
            self.numeric("366", f"#{name} :End of NAMES")
            return
        names = []
        for r in db.conn().execute(
            "SELECT u.username FROM channel_members cm JOIN users u ON u.id = cm.user_id "
            "WHERE cm.channel_id = ?",
            (ch["id"],),
        ).fetchall():
            names.append(r["username"])
        self.numeric("353", f"= #{ch['name']} :{' '.join(names)}")
        self.numeric("366", f"#{ch['name']} :End of NAMES")

    def _list(self):
        self.numeric("321", "Channel :Users  Name")
        if self.workspace:
            for r in db.conn().execute(
                "SELECT name, topic FROM channels WHERE workspace_id = ? AND is_dm = 0 AND is_archived = 0",
                (self.workspace["id"],),
            ).fetchall():
                self.numeric("322", f"#{r['name']} 0 :{r['topic']}")
        self.numeric("323", ":End of /LIST")

    async def _topic(self, chan_name: str, new_topic: Optional[str]):
        name = chan_name.lstrip("#").lower()
        if not self.workspace:
            return
        ch = _channel_by_name(self.workspace["id"], name)
        if not ch:
            self.numeric("403", f"#{name} :No such channel")
            return
        if new_topic is None:
            if ch["topic"]:
                self.numeric("332", f"#{ch['name']} :{ch['topic']}")
            else:
                self.numeric("331", f"#{ch['name']} :No topic is set")
            return
        try:
            db.conn().execute("BEGIN IMMEDIATE")
            db.conn().execute("UPDATE channels SET topic = ? WHERE id = ?", (new_topic[:250], ch["id"]))
            updated = db.conn().execute("SELECT * FROM channels WHERE id = ?", (ch["id"],)).fetchone()
            db.commit_event(ch["id"], "channel.updated", {"channel": util.public_channel(updated)})
            db.conn().execute("COMMIT")
        except Exception:
            db.tx_rollback()
        self.send(f"{self.prefix()} TOPIC #{ch['name']} :{new_topic[:250]}")

    async def _privmsg(self, target: str, body: str):
        if not target or not body:
            return
        if not target.startswith("#"):
            return
        if not self.workspace:
            self.numeric("403", f"{target} :No such channel")
            return
        ch = _channel_by_name(self.workspace["id"], target)
        if not ch:
            self.numeric("403", f"{target} :No such channel")
            return
        if ch["is_archived"]:
            self.numeric("404", f"{target} :Channel is archived")
            return
        # Insert message and event in a single transaction.
        try:
            db.conn().execute("BEGIN IMMEDIATE")
            mid = db.gen_id()
            now = db.now_ms()
            eid, seq = db.commit_event(ch["id"], "__placeholder__", {})
            db.conn().execute(
                "INSERT INTO messages(id, channel_id, author_id, body, parent_id, created_at, edited_at, deleted, seq, reply_count) "
                "VALUES (?, ?, ?, ?, NULL, ?, NULL, 0, ?, 0)",
                (mid, ch["id"], self.user["id"], body, now, seq),
            )
            mentions = util.resolve_mentions(ch["workspace_id"], ch["id"], body, self.user["id"])
            for uid in mentions:
                db.conn().execute(
                    "INSERT OR IGNORE INTO message_mentions(message_id, user_id) VALUES (?, ?)",
                    (mid, uid),
                )
            msg_row = db.conn().execute("SELECT * FROM messages WHERE id = ?", (mid,)).fetchone()
            obj = util.public_message(msg_row)
            db.conn().execute(
                "UPDATE events SET kind = ?, payload = ? WHERE id = ?",
                ("message.new", json.dumps({"message": obj}), eid),
            )
            db.conn().execute("COMMIT")
        except Exception:
            db.tx_rollback()
            self.numeric("404", f"{target} :Cannot send to channel")
            return


# --- Server ----------------------------------------------------------------

class IrcGateway:
    def __init__(self):
        self.sessions: list[IrcSession] = []
        self.last_event_id: int = 0

    def add(self, s: IrcSession):
        if s not in self.sessions:
            self.sessions.append(s)

    def remove(self, s: IrcSession):
        try:
            self.sessions.remove(s)
        except ValueError:
            pass

    async def fan_out_loop(self):
        while True:
            try:
                await asyncio.sleep(0.25)
                evts = db.fetch_events_after_id(self.last_event_id, limit=500)
                for e in evts:
                    if e["id"] <= self.last_event_id:
                        continue
                    self.last_event_id = e["id"]
                    if e["kind"] != "message.new":
                        continue
                    payload = e["payload"]
                    msg = payload.get("message")
                    if not msg:
                        continue
                    cid = e["channel_id"]
                    ch = db.conn().execute("SELECT * FROM channels WHERE id = ?", (cid,)).fetchone()
                    if not ch:
                        continue
                    author = msg.get("author", {})
                    nick = author.get("username", "huddle")
                    body = msg.get("body", "")
                    for s in list(self.sessions):
                        if not s.workspace or s.workspace["id"] != ch["workspace_id"]:
                            continue
                        if ch["name"] not in s.joined:
                            continue
                        if author.get("id") == (s.user["id"] if s.user else None):
                            # Don't echo back to the originating IRC user.
                            continue
                        for line in body.splitlines() or [""]:
                            s.send(f":{nick}!{nick}@huddle PRIVMSG #{ch['name']} :{line}")
            except asyncio.CancelledError:
                break
            except Exception:
                await asyncio.sleep(0.5)


async def run_irc():
    db.init_db()
    gw = IrcGateway()
    gw.last_event_id = db.latest_event_id()

    async def handle(reader, writer):
        sess = IrcSession(reader, writer, gw)
        await sess.run()

    server = await asyncio.start_server(handle, IRC_HOST, IRC_PORT, reuse_address=True)
    fan = asyncio.create_task(gw.fan_out_loop())
    try:
        async with server:
            await server.serve_forever()
    finally:
        fan.cancel()
        try:
            await fan
        except Exception:
            pass

File: artifacts/slack-clone/trial-3/source/irc.js
const net = require('net');
const WebSocket = require('ws');
const http = require('http');
const { getDb, generateId } = require('./lib/db');

const HUB_PORT = 9000;
const IRC_PORT = 6667;

const db = getDb();

const clients = new Map(); // socket -> { nick, userId, channels: Set }
const channelClients = new Map(); // #channelName -> Set<socket>
const channelIdToName = new Map();
const channelNameToId = new Map();

function refreshChannelMaps() {
  channelIdToName.clear();
  channelNameToId.clear();
  const channels = db.prepare('SELECT id, name FROM channels').all();
  for (const ch of channels) {
    channelIdToName.set(ch.id, ch.name);
    channelNameToId.set(ch.name, ch.id);
  }
}
refreshChannelMaps();

// Dedup for incoming events
const seenEvents = new Map();
function isDuplicate(event) {
  if (!event.channel_id || event.seq === undefined) return false;
  if (!seenEvents.has(event.channel_id)) seenEvents.set(event.channel_id, new Set());
  const s = seenEvents.get(event.channel_id);
  if (s.has(event.seq)) return true;
  s.add(event.seq);
  if (s.size > 5000) { const arr = Array.from(s); arr.sort((a,b)=>a-b); for (let i=0;i<arr.length-2500;i++) s.delete(arr[i]); }
  return false;
}

// Hub connection
let hubWs = null;
let hubConnected = false;

function connectHub() {
  try {
    hubWs = new WebSocket(`ws://127.0.0.1:${HUB_PORT}`);
    hubWs.on('open', () => { hubConnected = true; console.log('[irc] connected to hub'); });
    hubWs.on('message', (data) => {
      try {
        const event = JSON.parse(data.toString());
        if (event.type === 'welcome') return;
        if (!isDuplicate(event)) handleEvent(event);
      } catch (e) {}
    });
    hubWs.on('close', () => { hubConnected = false; setTimeout(connectHub, 2000); });
    hubWs.on('error', () => { hubConnected = false; setTimeout(connectHub, 2000); });
  } catch (e) { setTimeout(connectHub, 2000); }
}
connectHub();

function handleEvent(event) {
  if (!event.channel_id) return;
  const channelName = channelIdToName.get(event.channel_id);
  if (!channelName) { refreshChannelMaps(); return; }
  const chName = `#${channelName}`;
  switch (event.type) {
    case 'message.new':
    case 'message.reply': {
      const msg = event.message;
      const author = msg.author;
      const nick = author ? author.username : 'unknown';
      relayToChannel(chName, `:${nick}!~${nick}@localhost PRIVMSG ${chName} :${msg.body}`);
      break;
    }
    case 'message.edited': {
      const nick = event.message?.author?.username || 'unknown';
      relayToChannel(chName, `:${nick}!~${nick}@localhost NOTICE ${chName} :(edited) ${event.message.body}`);
      break;
    }
    case 'member.joined': {
      if (event.user && event.user.username) {
        relayToChannel(chName, `:${event.user.username}!~${event.user.username}@localhost JOIN ${chName}`);
      }
      break;
    }
  }
}

function relayToChannel(channelName, line) {
  if (channelClients.has(channelName)) {
    for (const sock of channelClients.get(channelName)) {
      sock.write(line + '\r\n');
    }
  }
}

function broadcastEvent(event) {
  // Send to hub
  if (hubConnected && hubWs && hubWs.readyState === 1) {
    hubWs.send(JSON.stringify(event));
  }
  // Also send directly to HTTP nodes
  const data = JSON.stringify(event);
  for (const port of [8000, 8001, 8002]) {
    const req = http.request({
      hostname: '127.0.0.1', port,
      path: '/__internal/event', method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) },
      timeout: 1000
    }, () => {});
    req.on('error', () => {});
    req.write(data);
    req.end();
  }
}

function nextSeq(channelId) {
  db.prepare('UPDATE channels SET seq_counter = seq_counter + 1 WHERE id = ?').run(channelId);
  const ch = db.prepare('SELECT seq_counter FROM channels WHERE id = ?').get(channelId);
  return ch.seq_counter;
}

// HTTP endpoint for internal events
const httpServer = http.createServer((req, res) => {
  if (req.method === 'POST' && req.url === '/__internal/event') {
    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', () => {
      try {
        const event = JSON.parse(body);
        if (!isDuplicate(event)) handleEvent(event);
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true }));
      } catch (e) { res.writeHead(400); res.end('{}'); }
    });
  } else {
    res.writeHead(200);
    res.end('irc ok');
  }
});
httpServer.listen(6700, '127.0.0.1', () => {
  console.log('[irc] http on 127.0.0.1:6700');
});

// TCP IRC server
const server = net.createServer((socket) => {
  let regState = { passed: false, nicked: false, usered: false, nick: '', userName: 'guest', realName: 'Guest', userId: null };
  let buffer = '';

  socket.on('data', (data) => {
    buffer += data.toString();
    const lines = buffer.split('\r\n');
    buffer = lines.pop() || '';
    for (const line of lines) {
      handleLine(socket, line.trim());
    }
  });

  function handleLine(sock, line) {
    if (!line) return;
    console.log(`[irc] recv: ${line}`);

    const parts = [];
    let current = '';
    let trailing = false;
    for (let i = 0; i < line.length; i++) {
      const ch = line[i];
      if (ch === ' ' && !trailing) {
        if (current) { parts.push(current); current = ''; }
      } else if (ch === ':' && (i === 0 || parts.length > 0 && current === '')) {
        trailing = true;
      } else {
        current += ch;
      }
    }
    if (current) parts.push(current);
    if (parts.length === 0) return;

    const command = parts[0].toUpperCase();

    // Handle pre-registration commands
    if (command === 'PASS') {
      const token = parts[1];
      const tok = db.prepare('SELECT * FROM tokens WHERE token = ?').get(token);
      if (!tok) {
        sock.write(':localhost 464 * :Password incorrect\r\n');
        sock.end();
        return;
      }
      regState.passed = true;
      regState.userId = tok.user_id;
      const user = db.prepare('SELECT * FROM users WHERE id = ?').get(tok.user_id);
      if (user) {
        regState.nick = user.username;
        regState.userName = user.username;
        regState.realName = user.display_name || user.username;
      }
      return;
    }

    if (command === 'NICK') {
      if (!regState.passed) {
        sock.write(':localhost 451 * :You have not registered\r\n');
        return;
      }
      const newNick = parts[1];
      // Check collision
      let collision = false;
      for (const [cs, info] of clients) {
        if (cs !== sock && info.nick === newNick) { collision = true; break; }
      }
      if (collision) {
        sock.write(`:localhost 433 * ${newNick} :Nickname is already in use\r\n`);
        return;
      }
      // If already registered, this is a nick change
      if (clients.has(sock)) {
        const info = clients.get(sock);
        const oldNick = info.nick;
        info.nick = newNick;
        for (const ch of info.channels) {
          relayToChannel(ch, `:${oldNick}!~${info.nick}@localhost NICK :${newNick}`);
        }
        return;
      }
      regState.nicked = true;
      regState.nick = newNick;
      // Check if we have both NICK and USER
      if (regState.nicked && regState.usered) {
        completeRegistration(sock);
      }
      return;
    }

    if (command === 'USER') {
      if (!regState.passed) {
        sock.write(':localhost 451 * :You have not registered\r\n');
        return;
      }
      if (parts.length >= 5) {
        regState.userName = parts[1];
        regState.realName = parts.slice(4).join(' ').replace(/^:/, '');
      }
      regState.usered = true;
      // Check if we have both NICK and USER
      if (regState.nicked && regState.usered) {
        completeRegistration(sock);
      }
      return;
    }

    // Unregistered clients
    if (!clients.has(sock)) {
      if (command === 'CAP' || command === 'PING') {
        if (command === 'PING') sock.write(`:localhost PONG localhost :${parts[1] || ''}\r\n`);
        return;
      }
      sock.write(':localhost 451 * :You have not registered\r\n');
      return;
    }

    const clientInfo = clients.get(sock);
    const nick = clientInfo.nick;
    const userId = clientInfo.userId;

    switch (command) {
      case 'JOIN': {
        const chName = parts[1];
        if (!chName || !chName.startsWith('#')) {
          sock.write(':localhost 403 * :No such channel\r\n');
          return;
        }
        const channelName = chName.slice(1);
        refreshChannelMaps();
        const channelId = channelNameToId.get(channelName);
        if (!channelId) {
          sock.write(`:localhost 403 ${nick} ${chName} :No such channel\r\n`);
          return;
        }
        const ch = db.prepare('SELECT * FROM channels WHERE id = ?').get(channelId);
        if (!ch) {
          sock.write(`:localhost 403 ${nick} ${chName} :No such channel\r\n`);
          return;
        }
        // Check membership
        const chMember = db.prepare('SELECT * FROM channel_members WHERE channel_id = ? AND user_id = ?').get(channelId, userId);
        if (!chMember) {
          sock.write(`:localhost 403 ${nick} ${chName} :You are not a member of this channel\r\n`);
          return;
        }
        // Join channel
        if (!channelClients.has(chName)) channelClients.set(chName, new Set());
        channelClients.get(chName).add(sock);
        clientInfo.channels.add(chName);

        const prefix = `:${nick}!~${clientInfo.userName || nick}@localhost`;
        sock.write(`${prefix} JOIN ${chName}\r\n`);
        relayToChannel(chName, `${prefix} JOIN ${chName}`);

        if (ch.topic) {
          sock.write(`:localhost 332 ${nick} ${chName} :${ch.topic}\r\n`);
        } else {
          sock.write(`:localhost 331 ${nick} ${chName} :No topic is set\r\n`);
        }

        const namesInChannel = [];
        for (const [cs, info] of clients) {
          if (info.channels.has(chName)) namesInChannel.push(info.nick);
        }
        sock.write(`:localhost 353 ${nick} = ${chName} :${namesInChannel.join(' ')}\r\n`);
        sock.write(`:localhost 366 ${nick} ${chName} :End of /NAMES list\r\n`);
        break;
      }

      case 'PRIVMSG': {
        const target = parts[1];
        const body = parts.slice(2).join(' ').replace(/^:/, '');
        if (!target || !body) {
          sock.write(`:localhost 411 ${nick} :No recipient given (PRIVMSG)\r\n`);
          return;
        }
        if (target.startsWith('#')) {
          const channelName = target.slice(1);
          refreshChannelMaps();
          const channelId = channelNameToId.get(channelName);
          if (!channelId) {
            sock.write(`:localhost 403 ${nick} ${target} :No such channel\r\n`);
            return;
          }
          const ch = db.prepare('SELECT * FROM channels WHERE id = ?').get(channelId);
          if (!ch || ch.is_archived) {
            sock.write(`:localhost 404 ${nick} ${target} :Cannot send to archived channel\r\n`);
            return;
          }

          const msgId = generateId();
          const now = new Date().toISOString();
          const seq = nextSeq(channelId);
          db.prepare('INSERT INTO messages (id, channel_id, author_id, body, created_at, seq) VALUES (?, ?, ?, ?, ?, ?)')
            .run(msgId, channelId, userId, body, now, seq);

          const author = db.prepare('SELECT * FROM users WHERE id = ?').get(userId);
          broadcastEvent({
            type: 'message.new', seq, channel_id: channelId,
            message: {
              id: msgId, channel_id: channelId, author_id: userId,
              author: { id: author.id, username: author.username, display_name: author.display_name || author.username, timezone: author.timezone, avatar_url: author.avatar_url, status_text: author.status_text, status_emoji: author.status_emoji },
              body, parent_id: null, reply_count: 0, created_at: now, edited_at: null, files: [], reactions: [], mentions: [], seq
            }
          });

          const prefix = `:${nick}!~${clientInfo.userName || nick}@localhost`;
          const relayMsg = `${prefix} PRIVMSG ${target} :${body}`;
          if (channelClients.has(target)) {
            for (const cs of channelClients.get(target)) {
              cs.write(relayMsg + '\r\n');
            }
          }
        } else {
          // DM - relay to target if connected
          for (const [cs, info] of clients) {
            if (info.nick === target) {
              cs.write(`:${nick}!~${clientInfo.userName || nick}@localhost PRIVMSG ${target} :${body}\r\n`);
              break;
            }
          }
        }
        break;
      }

      case 'PING':
        sock.write(`:localhost PONG localhost :${parts[1] || ''}\r\n`);
        break;

      case 'PONG':
        break;

      case 'QUIT':
        sock.write(':localhost ERROR :Closing link\r\n');
        sock.end();
        break;

      case 'NAMES': {
        const chName = parts[1];
        if (chName && chName.startsWith('#')) {
          const namesInChannel = [];
          if (channelClients.has(chName)) {
            for (const cs of channelClients.get(chName)) {
              const info = clients.get(cs);
              if (info) namesInChannel.push(info.nick);
            }
          }
          sock.write(`:localhost 353 ${nick} = ${chName} :${namesInChannel.join(' ')}\r\n`);
          sock.write(`:localhost 366 ${nick} ${chName} :End of /NAMES list\r\n`);
        }
        break;
      }

      case 'WHO': {
        const chName = parts[1];
        if (chName && chName.startsWith('#')) {
          if (channelClients.has(chName)) {
            for (const cs of channelClients.get(chName)) {
              const info = clients.get(cs);
              if (info) sock.write(`:localhost 352 ${nick} ${chName} ~${info.nick} localhost localhost ${info.nick} H :0 ${info.nick}\r\n`);
            }
          }
          sock.write(`:localhost 315 ${nick} ${chName} :End of WHO list\r\n`);
        }
        break;
      }

      case 'LIST':
        refreshChannelMaps();
        for (const [chName, chId] of channelNameToId) {
          const ch = db.prepare('SELECT * FROM channels WHERE id = ?').get(chId);
          if (ch && !ch.is_private) {
            const count = channelClients.has(`#${chName}`) ? channelClients.get(`#${chName}`).size : 0;
            sock.write(`:localhost 322 ${nick} #${chName} ${count} :${ch.topic || ''}\r\n`);
          }
        }
        sock.write(`:localhost 323 ${nick} :End of LIST\r\n`);
        break;

      case 'TOPIC': {
        const chName = parts[1];
        const newTopic = parts.slice(2).join(' ').replace(/^:/, '');
        if (!chName || !chName.startsWith('#')) {
          sock.write(`:localhost 403 ${nick} :No such channel\r\n`);
          return;
        }
        const channelName = chName.slice(1);
        refreshChannelMaps();
        const channelId = channelNameToId.get(channelName);
        if (!channelId) {
          sock.write(`:localhost 403 ${nick} ${chName} :No such channel\r\n`);
          return;
        }
        if (newTopic) {
          const ch = db.prepare('SELECT * FROM channels WHERE id = ?').get(channelId);
          if (ch && ch.workspace_id) {
            const ws = db.prepare('SELECT * FROM workspaces WHERE id = ?').get(ch.workspace_id);
            const member = db.prepare('SELECT * FROM workspace_members WHERE workspace_id = ? AND user_id = ?').get(ws.id, userId);
            if (member && ['owner', 'admin'].includes(member.role)) {
              db.prepare('UPDATE channels SET topic = ? WHERE id = ?').run(newTopic, channelId);
              relayToChannel(chName, `:${nick}!~${clientInfo.userName || nick}@localhost TOPIC ${chName} :${newTopic}`);
            } else {
              sock.write(`:localhost 482 ${nick} ${chName} :You're not channel operator\r\n`);
            }
          }
        } else {
          const ch = db.prepare('SELECT * FROM channels WHERE id = ?').get(channelId);
          if (ch && ch.topic) {
            sock.write(`:localhost 332 ${nick} ${chName} :${ch.topic}\r\n`);
          } else {
            sock.write(`:localhost 331 ${nick} ${chName} :No topic is set\r\n`);
          }
        }
        break;
      }

      case 'MODE':
        sock.write(`:localhost 221 ${nick} +\r\n`);
        break;

      case 'CAP':
        break;

      default:
        sock.write(`:localhost 421 ${nick} ${command} :Unknown command\r\n`);
        break;
    }
  }

  function completeRegistration(sock) {
    const info = {
      nick: regState.nick,
      userId: regState.userId,
      userName: regState.userName,
      realName: regState.realName,
      channels: new Set()
    };
    clients.set(sock, info);
    const n = info.nick;
    sock.write(`:localhost 001 ${n} :Welcome to Huddle IRC Gateway, ${n}\r\n`);
    sock.write(`:localhost 002 ${n} :Your host is localhost, running version 1.0\r\n`);
    sock.write(`:localhost 003 ${n} :This server was created now\r\n`);
    sock.write(`:localhost 004 ${n} localhost 1.0 o o\r\n`);
    sock.write(`:localhost 005 ${n} CHANTYPES=# PREFIX=(o)@ MODES= CHANLIMIT=#: NICKLEN=16 TOPICLEN=250 :are supported by this server\r\n`);
  }

  socket.on('close', () => {
    const info = clients.get(socket);
    if (info) {
      for (const chName of info.channels) {
        if (channelClients.has(chName)) {
          channelClients.get(chName).delete(socket);
          if (channelClients.get(chName).size === 0) channelClients.delete(chName);
        }
        relayToChannel(chName, `:${info.nick}!~${info.userName || info.nick}@localhost PART ${chName} :`);
      }
      clients.delete(socket);
    }
  });

  socket.on('error', () => { socket.destroy(); });
});

server.listen(IRC_PORT, '0.0.0.0', () => {
  console.log(`[irc] listening on 0.0.0.0:${IRC_PORT}`);
});
